### ISSUE TITLE
[VT-127] Bad trailing backslash handling in regex search and filter (patch attached)

### ISSUE THREAD

**Issue by taskwarrior** [scottkosty/vit#148]
_Rowan Thorpe on 2015-11-29T13:36:59Z says:_ 

 Using the version of Vit pulled from latest git-master, if a search is defined which ends in a backslash the entire vit process exits with exit-code 255, with the following error-message, and leaves the terminal in a jumbled state (invisible cursor, no carriage-return on newlines, etc - obviously the Curses lib doesn't get a chance to cleanup after itself):

```
Trailing \ in regex m/\/ at /usr/bin/vit line 1898.
```

and when a trailing backslash is specified in a filter prompt it doesn't kill vit, but the underlying taskwarrior-command outputs error-text which interpolates badly with the Curses prompt:

```
Filter: \Configuration override rc.defaultwidth:166
                                                   Configuration override rc.defaultheight:0
                                                                                            Configuration override rc._forcecolor:on
Error: task next \: no matches                                                                                                         Top
```

It seems members of the public can't register at the tasktools stash repo-management tool, so I have instead attached a git-diff of a patch which "Works For Me" (against latest Vit master) for these issues. The regex sanitization can be done in either of two ways (I used only one in the patch of course):

{code:title=either...}
if ( $report_lines[$i] =~ /\Q$search_pat\E/i ) {
{code}

{code:title=or...}
my $quoted_search_pat = quotemeta($search_pat);
if ( $report_lines[$i] =~ /$quoted_search_pat/i ) {
{code}

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2015-11-29T13:36:59Z
Modified: 2016-01-11T01:21:41Z
```

**Comment by taskwarrior** [comment]
_Scott Kostyshak on 2016-01-10T04:58:13Z says:_ 

 Thanks for this clean write-up of this issue and for the patch, Rowan! Here are my thoughts:
*  The issue at hand is more general than a trailing backslash I believe. It is about any invalid regular expression.
*  After thinking about this for a while, I think the expected behavior should be like in Vim: the search should accept regular expressions, but if a regular expression is invalid (like having a trailing backslash) then the string should be treated as a fixed string (i.e. the behavior after your patch). VIT should certainly no exit like it currently does if there is a trailing backslash.
*  The above behavior should be documented. Specifically, the help should state that '/' performs a regex search.
*  Was the substr trimming of the backslash intended? I think the rest of your patch is sufficient and better for the intended purpose.
*  I think the behavior for filter should be more like you implement in your patch: 
*  Although it is not the most common structure, master branch is not the development branch. Consistent with taskwarrior, master branch is the most recently released stable branch. 1.3 is currently the development branch.
*  I just made a patch that improves exiting with error from VIT (to avoid the jumbled state of the terminal that you note). So that is a separate bug that will be fixed thanks to your feedback.

**Comment by taskwarrior** [comment]
_Scott Kostyshak on 2016-01-11T01:21:41Z says:_ 

 I've made the fixes at e87c1232 (clean exit so terminal is usable) and 77f2eef0 (do not exit on invalid regex). Can you test the 1.3 branch?

Note also a couple of other similar fixes that you might be interested in. For example, annotations are always quoted now (see bba05dd6), so any character can be used in them.

**Comment by pbeckingham** [comment]
[VT-127_0001-Fix-trailing-backslash-handling-in-search-filter.patch.txt](https://github.com/scottkosty/vit/files/1755272/VT-127_0001-Fix-trailing-backslash-handling-in-search-filter.patch.txt)
